lib/ansible/module_utils/cisco_ucs.py

```diff
@@ -0,0 +1,89 @@
+# still belong to the author of the module, and may assign their own license
+# to the complete work.
+#
+# (c) 2016 Red Hat Inc.
+#
+# Redistribution and use in source and binary forms, with or without modification,
+# are permitted provided that the following conditions are met:
+#
+#    * Redistributions of source code must retain the above copyright
+#      notice, this list of conditions and the following disclaimer.
+#    * Redistributions in binary form must reproduce the above copyright notice,
+#      this list of conditions and the following disclaimer in the documentation
+#      and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+# IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+# USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+
+
+try:
+    import ucsmsdk
+    HAS_UCSMSDK = True
+except:
+    HAS_UCSMSDK = False
+
+try:
+    import ucsm_apis
+    HAS_UCSMAPI = True
+except:
+    HAS_UCSMAPI = False
+
+
+class UcsConnection():
+
+    def __init__(self, module):
+        if HAS_UCSMSDK is False:
+            results = {}
+            results["msg"] = "ucsmsdk is not installed"
+            module.fail_json(**results)
+        if HAS_UCSMAPI is False:
+            results = {}
+            results["msg"] = "ucsm_apis is not installed"
+            module.fail_json(**results)
+        self.module = module
+        self.handle = None
+
+    def login(self):
+        ansible = self.module.params
+        server = ansible.get('ucs_server')
+        if server:
+            return server
+
+        from ucsmsdk.ucshandle import UcsHandle
+        results = {}
+        try:
+            server = UcsHandle(ip=ansible["ucs_ip"],
+                               username=ansible["ucs_username"],
+                               password=ansible["ucs_password"],
+                               port=ansible["ucs_port"],
+                               secure=ansible["ucs_secure"],
+                               proxy=ansible["ucs_proxy"])
+            server.login()
+        except Exception as e:
+            results["msg"] = str(e)
+            self.module.fail_json(**results)
+        self.handle = server
+        return server
+
+    def logout(self):
+        server = self.module.params.get('ucs_server')
```

Comment:
So I would do this differently. We just designed a similar interface for Cisco ACI and to make the modules as minimal as possible, you can simply do the login automatically when you create the object. And this would ensure this object would already exist. within the object.

Also I don't think **ucs_server** is a good name for the purpose. Why not call it **login_handle** or something similar ?

Reply:
Other modules using a REST API have a parameter **token** which consists of a token (usually a hash or serialized input) needed to use an existing login. Your settings for UCS can easily be added using YAML anchors and references.

See: https://en.wikipedia.org/wiki/YAML

```yaml
- hosts: localhost
  vars:
    ucs_parameters: &ucs_server
      hostname: 1.2.3.4
      username: admin
      password: c1sco!123
      no_proxy: yes

- ucs_dns:
    <<: *ucs_server
    name: 1.2.3.4
    description: Internal DNS server

- ucs_hostname:
    <<: *ucs_server
    name: foobar.ucs.cisco.com
    description: Internal server name
```